contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,417 @@
+
+While Kubernetes has been dealing with these issues, the broader storage community has also been dealing with a fragmented story for how to make their storage system available in different Container Orchestration Systems (COs). Storage vendors have to either write and support multiple volume drivers for different COs or choose to not support some COs.
+
+The Container Storage Interface (CSI) is a specification that resulted from cooperation between community members from various COs--including Kubernetes, Mesos, Cloud Foundry, and Docker. The goal of this interface is to establish a standardized mechanism for COs to expose arbitrary storage systems to their containerized workloads.
+
+The primary motivation for Storage vendors to adopt the interface is a desire to make their system available to as many users as possible with as little work as possible. The primary motivation for COs to adopt the interface is to invest in a mechanism that will enable their users to use as many different storage systems as possible. In addition, for Kubernetes, adopting CSI will have the added benefit of moving volume plugins out of tree, and enabling volume plugins to be containerized.
+
+### Links
+
+* Container Storage Interface (CSI) Spec
+
+## Objective
+
+The objective of this document is to document all the requirements for enabling a CSI compliant volume plugin (a CSI volume driver) in Kubernetes.
+
+## Goals
+
+* Define Kubernetes API for interacting with an arbitrary, third-party CSI volume drivers.
+* Define mechanism by which Kubernetes master and node components will securely communicate with an arbitrary, third-party CSI volume drivers.
+* Define mechanism by which Kubernetes master and node components will discover and register an arbitrary, third-party CSI volume driver deployed on Kubernetes.
+* Recommend packaging requirements for Kubernetes compatible, third-party CSI Volume drivers.
+* Recommend deployment process for Kubernetes compatible, third-party CSI Volume drivers on a Kubernetes cluster.
+
+## Non-Goals
+* Replace [Flex Volume plugin](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md)
+  * The Flex volume plugin exists as an exec based mechanism to create “out-of-tree” volume plugins.
+  * Because Flex drivers exist and depend on the Flex interface, it will continue to be supported with a stable API.
+  * The CSI Volume plugin will co-exist with Flex volume plugin.
+
+## Design Overview
+
+To support CSI Compliant Volume plugins, a new in-tree CSI Volume plugin will be introduced in Kubernetes. This new volume plugin will be the mechanism by which Kubernetes users (application developers and cluster admins) interact with external CSI volume drivers.
+
+The mount/unmount calls for the new in-tree CSI volume plugin will directly invoke `NodePublishVolume` and `NodeUnpublishVolume` CSI RPCs through a unix domain socket on the node machine.
```

Comment:
MountDevice/UnmountDevice calls?

Reply:
Mount/Unmount should be triggered by the Setup() and TearDown on the VolumePlugin Impl of the plugin.

Reply:
Changed to SetUp and TearDown to be more precise.

Reply:
In one of the previous calls we wanted "NodePublish/Unpublish" to actually do "MountDevice & UnmountDevice" so that we call mount only once per volume per node. @saad-ali would know the final conclusion on it.

Reply:
Yep, tracking that in https://github.com/container-storage-interface/spec/issues/119